### ISSUE TITLE
IDE0052 is incorrectly reported for a member referenced by nameof(...)

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#65379]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/IDE0052-is-incorrectly-reported-for-a-me/10193399)._

---
[severity:It's more difficult to complete my work] [regression] [worked-in:17.3]
Have this NUnit test class:

```
[TestFixture]
public class TestClass {
  private readonly int[] values; // IDE0052

  [Test]
  [TestCaseSource(nameof(values))]
  public void TestMethod(int[] values) {
     /// bla bla
  }
}
```

VS 17.4 reports that "values" is not referred and can be removed despite that `nameof` references it.

---
#### Feedback Bot on 11/8/2022, 04:55 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by Youssef1313** [comment]
The warning looks correct to me. `nameof(values)` refers to the `values` parameter, not the `values` field. The behavior change is caused by `nameof(parameter)` feature.

**Comment by cremor** [comment]
@Youssef1313 Is this dependend on the C# version which is in use?
So with C# versions before 11 IDE0052 isn't shown?

**Comment by Youssef1313** [comment]
@cremor The feature doesn't seem gated under a language version. So the behavior depends on the *compiler* version. I think you should be fine by choosing different names for the parameter and the field. My preference is to use `_values` for the field (which is the well-known .NET naming convention) and keep `values` for the parameter name. You then should use `nameof(_values)`

**Comment by cremor** [comment]
I'm not the reporter of this issue, and I'm also using the underscore naming convention for fields, so I'm not affected. I'm just interested in this issue.

The reporter of this issue also reported another similar one: https://developercommunity.visualstudio.com/t/Inline-rename-incorrectly-renames-unrela/10193406
There he wrote that the project targets .NET 4.8, so no C# 11.

Yes, both issues can be prevented by having a better naming convention. But both issues can also lead to hard to find bugs for code that was perfectly fine pre 17.4.

**Comment by Youssef1313** [comment]
@cremor `nameof(parameter)` will work on any C# version as long as the compiler version includes it. So it doesn't matter whether the project is .NET Framework 4.8 or is using older C# version.

The same applies to the reported rename issue.

**Comment by cremor** [comment]
Oh, I see. Thanks for explaining that. I didn't expect a new language version to work in old C# versions.

**Comment by ssg** [comment]
Yes, this is .NET Framework 4.8.1. This issue didn't happen with NetAnalyzers 6.0, but VS 2022 deprecated it with compiler warnings, so we had to remove the package and use what's in VS itself. I pushed a repro at https://github.com/ssg/IDE0052Repro. This is the code:

```csharp
    [TestFixture]
    public class ReproTest
    {
        private static readonly int[] values = new int[] { -1, -2, -3 };

        [Test]
        [TestCaseSource(nameof(values))]
        public void Test(int values)
        {
            Assert.That(values, Is.Negative);
        }
    }
```

The compiler says the private static member is unused:

> IDE0052	Private member 'ReproTest.values' can be removed as the value assigned to it is never read

But NUnit is able to find the class member using it:

<img width="276" alt="image" src="https://user-images.githubusercontent.com/241217/201735764-f62f042c-e372-4302-bd5b-3827492e26e9.png">

As I understand, NUnit and Roslyn have differing opinions on what `nameof` refers to, and it's just hardcoded in NUnit that the name parameter to `TestCaseSource` attribute always refers to a class member which prevents things from breaking. So, it looks like there's nothing to do other than avoiding naming clashes with arguments and members here.

**Comment by Youssef1313** [comment]
@ssg NUnit doesn't care whether you use `nameof` or not in the first place. At the end, this is just a string constant. So, even `[TestCaseSource("values")]` will work (but I don't recommend hard-coding the string - just demonstrating NUnit's view of things).

**Comment by Youssef1313** [comment]
If you're dealing with the compiler APIs, you get to see what the thing inside `nameof` binds to (as a symbol), but outside the compiler, it's just a string.

So, the analyzer sees that `values` inside `nameof` binds to the parameter (because of the new feature). Hence, the field `values` is considered unused.

**Comment by ssg** [comment]
@Youssef1313 I'm not surprised that compiler knows about it, but do you know if, say, source generators also have access to that information?

**Comment by Youssef1313** [comment]
@ssg Generators deal with the compiler API, so I guess nothing blocks them from knowing that info.
